File: api/app/schemas/todo.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Literal
from datetime import date, datetime
import re


# ============================================================================
# TODO Item Schemas
# ============================================================================

class LinkedTaskInfo(BaseModel):
    """Read-only summary information for linked tasks/subtasks"""
    id: str
    title: str
    status: str
    due_date: Optional[date] = None
    assigned_to: Optional[str] = None
    parent_id: Optional[str] = None  # user_story_id for tasks, task_id for subtasks

    class Config:
        from_attributes = True


class TodoItemBase(BaseModel):
    """Base schema for TODO items with common fields"""
    title: str = Field(..., min_length=1, max_length=255, description="Title of the TODO item")
    description: Optional[str] = Field(None, max_length=2000, description="Detailed description")
    target_date: date = Field(..., description="Target date for the TODO item")
    visibility: Literal["public", "private"] = Field("private", description="Visibility setting")

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        """Validate title is not empty after stripping whitespace"""
        if not v or not v.strip():
            raise ValueError("Title cannot be empty")
        return v.strip()


class TodoItemCreate(TodoItemBase):
    """Schema for creating a new TODO item"""
    linked_entity_type: Optional[Literal["task", "subtask"]] = Field(
        None, 
        description="Type of linked entity"
    )
    linked_entity_id: Optional[str] = Field(
        None, 
        max_length=20,
        description="ID of the linked task or subtask"
    )

    @field_validator("linked_entity_id")
    @classmethod
    def validate_linked_entity(cls, v: Optional[str], info) -> Optional[str]:
        """Validate that if linked_entity_id is provided, linked_entity_type must also be provided"""
        if v is not None:
            # Access linked_entity_type from the data being validated
            linked_entity_type = info.data.get("linked_entity_type")
            if not linked_entity_type:
                raise ValueError("linked_entity_type must be provided when linked_entity_id is set")
        return v


class TodoItemUpdate(BaseModel):
    """Schema for updating an existing TODO item"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    target_date: Optional[date] = None
    visibility: Optional[Literal["public", "private"]] = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: Optional[str]) -> Optional[str]:
        """Validate title is not empty after stripping whitespace"""
        if v is not None:
            if not v.strip():
                raise ValueError("Title cannot be empty")
            return v.strip()
        return v


class TodoItemResponse(TodoItemBase):
    """Schema for TODO item responses"""
    id: str
    user_id: str
    linked_entity_type: Optional[Literal["task", "subtask"]] = None
    linked_entity_id: Optional[str] = None
    linked_entity_info: Optional[LinkedTaskInfo] = None
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TodoItemList(BaseModel):
    """Schema for paginated list of TODO items"""
    items: List[TodoItemResponse]
    total: int


class MoveTodoItemRequest(BaseModel):
    """Schema for moving a TODO item to a different date"""
    target_date: date = Field(..., description="New target date for the TODO item")


class LinkTodoItemRequest(BaseModel):
    """Schema for linking a TODO item to a task or subtask"""
    entity_type: Literal["task", "subtask"] = Field(..., description="Type of entity to link")
    entity_id: str = Field(..., max_length=20, description="ID of the entity to link")


# ============================================================================
# ADHOC Note Schemas
# ============================================================================

class AdhocNoteBase(BaseModel):
    """Base schema for ADHOC notes with common fields"""
    title: str = Field(..., min_length=1, max_length=255, description="Title of the note")
    content: Optional[str] = Field(None, description="Content of the note")
    color: str = Field("#FFEB3B", pattern=r"^#[0-9A-Fa-f]{6}$", description="Hex color code")

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        """Validate title is not empty after stripping whitespace"""
        if not v or not v.strip():
            raise ValueError("Title cannot be empty")
        return v.strip()

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str) -> str:
        """Validate color is a valid hex color code"""
        if not re.match(r"^#[0-9A-Fa-f]{6}$", v):
            raise ValueError("Color must be a valid hex color code (e.g., #FFEB3B)")
        return v.upper()


class AdhocNoteCreate(AdhocNoteBase):
    """Schema for creating a new ADHOC note"""
    pass


class AdhocNoteUpdate(BaseModel):
    """Schema for updating an existing ADHOC note"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    content: Optional[str] = None
    color: Optional[str] = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: Optional[str]) -> Optional[str]:
        """Validate title is not empty after stripping whitespace"""
        if v is not None:
            if not v.strip():
                raise ValueError("Title cannot be empty")
            return v.strip()
        return v

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: Optional[str]) -> Optional[str]:
        """Validate color is a valid hex color code"""
        if v is not None:
            if not re.match(r"^#[0-9A-Fa-f]{6}$", v):
                raise ValueError("Color must be a valid hex color code (e.g., #FFEB3B)")
            return v.upper()
        return v


class AdhocNoteResponse(AdhocNoteBase):
    """Schema for ADHOC note responses"""
    id: str
    user_id: str
    position: int
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AdhocNoteList(BaseModel):
    """Schema for list of ADHOC notes"""
    notes: List[AdhocNoteResponse]
    total: int


class ReorderAdhocNoteRequest(BaseModel):
    """Schema for reordering an ADHOC note"""
    position: int = Field(..., ge=0, description="New position for the note (must be >= 0)")

    @field_validator("position")
    @classmethod
    def validate_position(cls, v: int) -> int:
        """Validate position is non-negative"""
        if v < 0:
            raise ValueError("Position must be greater than or equal to 0")
        return v

File: ui/src/types/todo.ts
/**
 * TypeScript type definitions for TODO feature
 */

// Base TODO entity interface
export interface BaseTodoEntity {
  id: string
  user_id: string
  is_deleted: boolean
  created_at: string
  updated_at: string
}

// TODO Item
export interface TodoItem extends BaseTodoEntity {
  title: string
  description?: string
  target_date: string // ISO date string (YYYY-MM-DD)
  visibility: 'public' | 'private'
  linked_entity_type?: 'task' | 'subtask'
  linked_entity_id?: string
  linked_entity_info?: LinkedTaskInfo
}

// Linked Task/Subtask Info (read-only summary)
export interface LinkedTaskInfo {
  id: string
  title: string
  status: string
  due_date?: string
  assigned_to?: string
  parent_id?: string // user_story_id for tasks, task_id for subtasks
}

// ADHOC Note
export interface AdhocNote extends BaseTodoEntity {
  title: string
  content?: string
  position: number
  color: string // Hex color code (#RRGGBB)
}

// Time Pane (for organizing TODO items by date)
export interface Pane {
  label: 'Yesterday' | 'Today' | 'Tomorrow' | 'Day After Tomorrow'
  date: Date
  dateString: string // ISO date string for API calls
  items: TodoItem[]
}

// API Request Types

export interface CreateTodoItemRequest {
  title: string
  description?: string
  target_date: string // ISO date string
  visibility: 'public' | 'private'
  linked_entity_type?: 'task' | 'subtask'
  linked_entity_id?: string
}

export interface UpdateTodoItemRequest {
  title?: string
  description?: string
  target_date?: string
  visibility?: 'public' | 'private'
}

export interface MoveTodoItemRequest {
  target_date: string // ISO date string
}

export interface LinkTodoItemRequest {
  entity_type: 'task' | 'subtask'
  entity_id: string
}

export interface CreateAdhocNoteRequest {
  title: string
  content?: string
  color?: string
}

export interface UpdateAdhocNoteRequest {
  title?: string
  content?: string
  color?: string
}

export interface ReorderAdhocNoteRequest {
  position: number
}

// API Response Types

export interface TodoItemsResponse {
  items: TodoItem[]
  total: number
}

export interface AdhocNotesResponse {
  notes: AdhocNote[]
  total: number
}

export interface TaskSummaryResponse extends LinkedTaskInfo {
  user_story_id?: string // For tasks
}

export interface SubtaskSummaryResponse extends LinkedTaskInfo {
  task_id?: string // For subtasks
}

// Query Parameters

export interface FetchTodoItemsParams {
  start_date?: string // ISO date string
  end_date?: string // ISO date string
  include_public?: boolean
}

// Form Data Types

export interface TodoItemFormData {
  title: string
  description?: string
  target_date: string
  visibility: 'public' | 'private'
  linked_entity_type?: 'task' | 'subtask'
  linked_entity_id?: string
}

export interface AdhocNoteFormData {
  title: string
  content?: string
  color?: string
}

// Constants

export const TODO_VISIBILITY_OPTIONS = ['public', 'private'] as const
export const TODO_ENTITY_TYPES = ['task', 'subtask'] as const
export const PANE_LABELS = ['Yesterday', 'Today', 'Tomorrow', 'Day After Tomorrow'] as const

// Default colors for ADHOC notes
export const ADHOC_NOTE_COLORS = [
  '#FFEB3B', // Yellow (default)
  '#FFC107', // Amber
  '#FF9800', // Orange
  '#FF5722', // Deep Orange
  '#E91E63', // Pink
  '#9C27B0', // Purple
  '#673AB7', // Deep Purple
  '#3F51B5', // Indigo
  '#2196F3', // Blue
  '#03A9F4', // Light Blue
  '#00BCD4', // Cyan
  '#009688', // Teal
  '#4CAF50', // Green
  '#8BC34A', // Light Green
  '#CDDC39', // Lime
] as const

// Type guards

export function isTodoItem(item: any): item is TodoItem {
  return (
    item &&
    typeof item === 'object' &&
    'title' in item &&
    'target_date' in item &&
    'visibility' in item
  )
}

export function isAdhocNote(note: any): note is AdhocNote {
  return (
    note &&
    typeof note === 'object' &&
    'title' in note &&
    'position' in note &&
    'color' in note
  )
}

export function isLinkedTodoItem(item: TodoItem): boolean {
  return !!(item.linked_entity_type && item.linked_entity_id)
}

// Utility types

export type TodoVisibility = typeof TODO_VISIBILITY_OPTIONS[number]
export type TodoEntityType = typeof TODO_ENTITY_TYPES[number]
export type PaneLabel = typeof PANE_LABELS[number]
export type AdhocNoteColor = typeof ADHOC_NOTE_COLORS[number]
